src/Resolvers/DomainTenantResolver.php

```diff
@@ -28,18 +28,34 @@ class DomainTenantResolver extends Contracts\CachedTenantResolver
 
     public function resolveWithoutCache(...$args): Tenant
     {
-        /** @var Domain $domain */
-        $domain = config('tenancy.domain_model')::where('domain', $args[0])->first();
+        $domain = $args[0];
 
-        if ($domain) {
-            static::$currentDomain = $domain;
+        /** @var Tenant|null $tenant */
+        $tenant = config('tenancy.tenant_model')::query()
+            ->whereHas('domains', function ($query) use ($domain) {
+                $query->select(['tenant_id', 'domain'])->where('domain', $domain);
+            })
+            ->with([
+                'domains' => function ($query) use ($domain) {
+                    $query->where('domain', $domain);
+                },
+            ])
+            ->first();
 
-            return $domain->tenant;
+        if ($tenant) {
+            static::$currentDomain = $tenant->domains->first();
+
+            return $tenant;
         }
 
         throw new TenantCouldNotBeIdentifiedOnDomainException($args[0]);
     }
 
+    public function tenantIdentifiedFromCache(Tenant $tenant, ...$args): void
+    {
+        static::$currentDomain = $tenant->domains->first();
```

Comment:
As I mentioned, you cannot use the first domain here. **Tenants can have multiple domains**. See what I suggested in the original thread.

Reply:
No, in this context you can, because the $tenant comes from DomainTenantResolver, and you can have multiple domains, but the **$tenant will have only the current domain loaded on the relationship**. Look at DomainTenantResolver line 35, where $tenant is fetched from database.

Reply:
[Explanation](https://github.com/stancl/tenancy/pull/633#discussion_r611895731).

Reply:
Okay, so the tenant instance will have incorrect data loaded in the $domains property? What if someone wants to use tenant()->domains to fetch all domains?

Reply:
You can fetch them from database with `tenant()->domains()->get()`, but I get your point on this.
So, you want me to refactor the code, so `tenant()->domains` still contains all domains, and when using cache, there are no calls to database, right? (I guess, when setting current domain, find it in `$tenant->domains->where('domain', $domain)`)
I am just trying to get rid of unnecessary calls to central database, because, the suggested method queries the domain from DB, but it is not necessary, as the Domain model already exists in cache.